MarkBind/markbind
# Implement peek feature for panels

**What is the purpose of this pull request?**


This PR resolves #905. 

**Overview of changes:**
Introduced various variables and directives to implement the 'peek' feature for panels.

**Anything you'd like to highlight / discuss:**
I'm quite unsure of the best approach to implement this feature. For my approach, I decided to make localExpanded to always have a true value whenever the panel is on show-preview mode. Thus, the content of the panel will always be displayed (effectively overriding the v-show). I then use css to achieve the effect of a 'peek' mode. To be specific, I used css transition, max-height, and overflow hidden to toggle between collapsed ('peek' mode) and expanded panel. 

I also tried to make the behaviour of the 'peek' mode panel as separate from the original panel's toggling behaviour as much as possible (as can be seen from the conditions used for the click behaviours, e.g. "@click.stop="shouldShowPreview ? closePreviewCard() : close()", this is to prevent conflicts between the original panel behaviours from the peek mode panel behaviours). 

However, I do face various issues in terms of nested panels. As my css transition relies heavily on max-height, we always need to know the expanded size of a panel before we can apply the transition smoothly (the default requirement for css transition for max-height). In nested panels, we have many unexpanded panels. Suppose there are multiple nested panels, it is difficult to set an **accurate** max-height for the panel we are concerned with. For example, if we look at the first layer panel, it can adopt different heights (depending on how many nested panels we expand). Thus, it is difficult to set an accurate max-height for css transition to achieve a smooth transition (the max-height has to be as accurate as possible for the transition to not be 'jerky'). 

At the moment, nested panels can be quite buggy in terms of transitions. But the 'peek' functionality works fine. 

Beyond the nested panels issue, I think generally it works alright. However, more tests have to be conducted to be sure. 

It would be great if there are some suggestions on how I can improve the implementation (can be in terms of code quality or even a better way of approaching the problem), or resolve the issue regarding nested panels. 

For now, I will continue to see how I can resolve the issue and think of a better way to implement this feature :-)

@ang-zeyu would you have any advice for me? 

**Testing instructions:**
`npm run test`

Using the existing user guide, add the 'show-preview' attribute to any panels and serve the program, you should be able to see the preview mode that looks something like below. 

![Screen Shot 2020-12-15 at 12 38 38 AM](https://user-images.githubusercontent.com/37920115/102108433-e3d9ef00-3e6d-11eb-98e7-fdbdb37c9724.png)



**Proposed commit message: (wrap lines at 72 characters)**
Implement peek feature for panels

To find out what content a panel contains, a user has to expand it. 
This may be inconvenient when the user just wants to have an idea 
of whether the panel contains relevant content. 

Let's implement a feature to let users have a glimpse of the content, 
which would help them decide whether it is worth to expand the panel. 

---

**Checklist:** :ballot_box_with_check:
